### PR TITLE
feat(UI): adds Header component

### PR DESCRIPTION
**Pull request by damienomurchu**
## Motivation
Add header to landing page screen. AEROGEAR-8375

## What
Header for landing page ui screen

## Why
header

## How
Create header component with PF4 & add to Landing Page

## Verification Steps
- `npm start` 
- visually confirm header is rendered on landing page, and includes avatar icon, username, and dropdown with logout option
- verify dropdown toggles opens and closed
- verify dropdown closes after logout option selected
- run `npm test` & confirm all tests pass

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

### CODE REVIEW

**Comment by damienomurchu** [comment]
@laurafitzgerald just pinging you on this to get your eyes as I try construct a header with PF4 components

File: ui/package.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
   "dependencies": {
     "@patternfly/react-core": "2.1.4",
     "@patternfly/react-table": "1.0.7",
+    "prop-types": "^15.7.2",
```
Comment (laurafitzgerald) [inline ui/package.json]: @damienomurchu I'm also using Prop Types here: https://github.com/aerogear/mobile-security-service/blob/master/ui/src/containers/AppsTableContainer.js#L3 but I didn't need to add it as a dependency. It may already be there as part of the React or Redux dependencies?

File: ui/src/components/common/Header.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    /* const logoProps = {
+      href: 'https://patternfly.org',
+      onClick: () => console.log('clicked logo'),
+      target: '_blank'
+    }; */
+
+
+    const toolbar = (
+      <>
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]: I'm not familiar with the use of `<>`. Is this a PF thing?

Comment (damienomurchu) [reply ui/src/components/common/Header.js]: A React thing actually - it's shorthand for [`<React.Fragment>`](https://reactjs.org/docs/fragments.html). Must see if it's actually needed here though

**Comment by damienomurchu** [comment]
thanks for the feedback @laurafitzgerald - if you get a chance could you take a look at the rendered header & see if it's what we were thinking of?

File: ui/src/components/common/Header.js (hunk 2 above)
Comment (laurafitzgerald) [reply ui/src/components/common/Header.js]: @damienomurchu given that you both open and close the Fragment with Toolbar, I'm pretty sure that you don't need it.

**Comment by laurafitzgerald** [comment]
> thanks for the feedback @laurafitzgerald - if you get a chance could you take a look at the rendered header & see if it's what we were thinking of?

@damienomurchu yes I would say it's on the right track. It's starting to look similar to the one in MDC. The black background looks well.

File: ui/src/components/common/Header.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+              />
+            </ToolbarItem>
+          </ToolbarGroup>
+        </Toolbar>
+      </>
+    );
+
+    const Header = (
+      <PageHeader
+        logo="Mobile Developer Console" logoProps={logoProps} toolbar={toolbar} />
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]:
```suggested
logo="Mobile Security Service" logoProps={logoProps} toolbar={toolbar} />
```

**Comment by damienomurchu** [comment]
@laurafitzgerald thanks - just looking at the flat, think just an avatar/ user icon next to the dropdown is required, and will add this. 

Will add some tests & ping you for re-review once I do so and remove the draft label

**Comment by damienomurchu** [comment]
Removing this from draft status & reviews welcome

**Comment by laurafitzgerald** [comment]
@damienomurchu 👀

File: ui/package.json (hunk 1 above)
Comment (damienomurchu) [reply ui/package.json]: removed now :+1:

**Comment by damienomurchu** [comment]
@laurafitzgerald thanks for your eyes. As discussed need to wire this up to the redux store now, and will ping you for review then.

File: ui/src/components/common/Header.css [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.mssHeader {
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply ui/src/components/common/Header.css]: Good question, and given that this was only one line of CSS, I was even considering adding it via an inline style to the component, and not having a seperate css file for the Header. I think can chat about this off-PR & not let it block the PR as if we can refactor easily even if this is merged in as it is above

File: ui/src/components/common/Header.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  onTitleClick = () => {
+    const { history } = this.props;
+    history.push('/');
+  };
+
+  onUserDropdownToggle(isUserDropdownOpen) {
+    this.setState({
+      isUserDropdownOpen
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: Added this to redux now. FWIW I don't like the idea of mixing UI state between redux & using setState, and think Redux should be the single place we manage state. We can discuss this off-PR though I think :)

File: ui/src/components/common/Header.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                  Log out
+                </DropdownItem>
+              ]}
+            />
+          </ToolbarItem>
+        </ToolbarGroup>
+      </Toolbar>
+    );
+
+    const Header = <PageHeader logo="Mobile Security Service" logoProps={logoProps} toolbar={toolbar} />;
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: @laurafitzgerald Yes, this is how the text is rendered in PF4. Will correct case to all uppercase

File: ui/src/components/common/Header.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    this.onTitleClick = this.onTitleClick.bind(this);
+    //this.onLogoutUser = this.onLogoutUser.bind(this);
+
+    this.onUserDropdownToggle = this.onUserDropdownToggle.bind(this);
+    this.onUserDropdownSelect = this.onUserDropdownSelect.bind(this);
+  }
+
+  onTitleClick = () => {
+    const { history } = this.props;
+    history.push('/');
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: @laurafitzgerald I agree, and have removed this for now. I think we should discuss how we intend to manage page history in our app (or if we choose to manage it at all)

File: ui/src/components/common/Header.js [hunk 8]
```diff
@@ -14,43 +14,28 @@ import { UserIcon } from '@patternfly/react-icons';
...
-    };
-  }
-
-  onTitleClick = () => {
-    console.log('onTitleClick()');
-    // const { history } = this.props;
-    // history.push('/');
+  onUserDropdownToggle = () => {
+    const isUserDropdownOpen = this.props.isUserDropdownOpen;
+    this.props.toggleHeaderDropdown(isUserDropdownOpen);
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]: @damienomurchu I did the same when I initially set up the `isappsrequestfailed` where I passed the value through the component, action and reducer. As you are reversing it in the reducer you don't need to pass the value down.

File: ui/src/components/common/Header.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (laurafitzgerald) [reply ui/src/components/common/Header.css]: Yes agreed, it's probably a bigger conversation. Happy to move the convo elsewhere to unblock this pr.

File: ui/src/components/common/Header.js (hunk 8 above)
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: good point @laurafitzgerald :+1:

File: ui/src/components/common/Header.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (laurafitzgerald) [reply ui/src/components/common/Header.js]: 👍 Could you propose a new task via the team channel? We could address it as part of either UI story in my opinion.

**Comment by laurafitzgerald** [comment]
@damienomurchu thanks for pushing the changes. There's one console.log in the reducer which can be removed. Also I'm not sure about the `MOBILE SECURITY SERVICE` being rendered in link format. i.e. in blue and underline. It's deviating from the MDC layout and the flats somewhat. Could be worth a chat with Chris about it if it's the only way to do it in PF4.

**Comment by damienomurchu** [comment]
> @damienomurchu thanks for pushing the changes. There's one console.log in the reducer which can be removed. Also I'm not sure about the `MOBILE SECURITY SERVICE` being rendered in link format. i.e. in blue and underline. It's deviating from the MDC layout and the flats somewhat. Could be worth a chat with Chris about it if it's the only way to do it in PF4.

Console.log removed now :)

Re `MOBILE SECURITY SERVICE` being rendered in link format, my line of thought is that this would enable the user to always navigate back to the root page of the UI and supplement our app navigation. Wdyt? Perhaps we can discuss this & the navigation in general in the UI?

**Comment by laurafitzgerald** [comment]
> > @damienomurchu thanks for pushing the changes. There's one console.log in the reducer which can be removed. Also I'm not sure about the `MOBILE SECURITY SERVICE` being rendered in link format. i.e. in blue and underline. It's deviating from the MDC layout and the flats somewhat. Could be worth a chat with Chris about it if it's the only way to do it in PF4.
> 
> Console.log removed now :)
> 
> Re `MOBILE SECURITY SERVICE` being rendered in link format, my line of thought is that this would enable the user to always navigate back to the root page of the UI and supplement our app navigation. Wdyt? Perhaps we can discuss this & the navigation in general in the UI?

Yes, I think it should be a home link. but I don't think it should behave like a normal link. i.e. it shouldn't be Blue and underlined. I think it should appear in the same way it does in the MDC and in the flats.

**Comment by damienomurchu** [comment]
@laurafitzgerald agree, that can be fixed :+1:

File: ui/src/components/common/Header.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import { css } from '@patternfly/react-styles';
+import './Header.css';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { toggleHeaderDropdown } from '../../actions/actions-ui';
+
+class Header extends React.Component {
+  onTitleClick = () => {};
+
+  onUserDropdownToggle = () => {
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]: @damienomurchu Looks like this method is a duplicate of `onUserDropdownSelect()`

File: ui/src/components/common/Header.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      <Toolbar>
+        <ToolbarGroup className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnLg)}>
+          <UserIcon />
+          <ToolbarItem className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnMd)}>
+            <Dropdown
+              isPlain
+              position="right"
+              onSelect={this.onUserDropdownSelect}
+              isOpen={this.props.isUserDropdownOpen}
+              toggle={<DropdownToggle onToggle={this.onUserDropdownToggle}>currentUserName</DropdownToggle>}
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]: @damienomurchu we'll probably do the implemenation later, but could you move the `currentUserName` hardcoded value to Redux state?

File: ui/src/components/common/Header.js (hunk 9 above)
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: @laurafitzgerald both needed - `onUserDropdownToggle` will be triggered when the unexpanded dropdown is clicked on, and `onUserDropdownSelect` is needed to close the dropdown when a dropdown item has been selected

File: ui/src/components/common/Header.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    const toolbar = (
+      <Toolbar>
+        <ToolbarGroup className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnLg)}>
+          <UserIcon />
+          <ToolbarItem className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnMd)}>
+            <Dropdown
+              isPlain
+              position="right"
+              onSelect={this.onUserDropdownSelect}
+              isOpen={this.props.isUserDropdownOpen}
```
Comment (laurafitzgerald) [inline ui/src/components/common/Header.js]: Hey, I was thinking about this one yesterday. Because both methods do the same thing, you could pass the same method to the isOpen for the Dropdown component and to the onToggle of the DropdownToggle component. wdyt?

Comment (damienomurchu) [reply ui/src/components/common/Header.js]: @laurafitzgerald was thinking something similar, to reduce code duplication & make any future changes easier :+1:

**Comment by damienomurchu** [comment]
@laurafitzgerald good point - will add to Detailed Apps UI page. Re redux testing, I was thinking about that too, and will do it in another PR I think - will also ping you probably re redux testing as its not something I've done any of yet :)

**Comment by laurafitzgerald** [comment]
@damienomurchu cool. There's some example of how the tests can be done in my [previous pr](https://github.com/aerogear/mobile-security-service/pull/75/files#diff-8ce8f218c3acf4b3d431453881378f99)

**Comment by damienomurchu** [comment]
@laurafitzgerald added redux test now for dropdown state if you could re-review (& approve if happy). Don't see need to add redux test for currentUsername until we implement it properly by the way, so have not added any redux test for that

File: ui/src/components/common/Header.js (hunk 11 above)
Comment (laurafitzgerald) [reply ui/src/components/common/Header.js]: @damienomurchu This one still needs to be done?

**Comment by damienomurchu** [comment]
Accidentally broke tests before, but fixed now

File: ui/src/reducers/index.js [hunk 12]
```diff
@@ -35,7 +43,7 @@ export default (state = initialState, action) => {
       };
     case APPS_SUCCESS:
       var fetchedApps = [];
-      action.result.forEach(app => {
+      action.result.forEach((app) => {
```
Comment (laurafitzgerald) [inline ui/src/reducers/index.js]: What's the reason for this change?

Comment (damienomurchu) [reply ui/src/reducers/index.js]: just the auto-linter, no functional change

File: ui/src/components/common/Header.js (hunk 11 above)
Comment (damienomurchu) [reply ui/src/components/common/Header.js]: @laurafitzgerald misread your previous comment - isOpen is just a boolean value, and onSelect is the event handler.

**Comment by damienomurchu** [comment]
@laurafitzgerald thanks - responded inline. If you're happy let me know & I'll merge this in

File: ui/src/components/common/Header.js (hunk 11 above)
Comment (laurafitzgerald) [reply ui/src/components/common/Header.js]: Sorry I mean onSelect and onToggle

Comment (damienomurchu) [reply ui/src/components/common/Header.js]: onToggle opens the dropdown, and onSelect closes it. I think its probably best to keep both methods to keep table implementation as close to standard configuration as possible though.

Comment (laurafitzgerald) [reply ui/src/components/common/Header.js]: I disagree. We aren't changing the the table implementation, only the methods that are passed to each of the required props. But it's a small detail, as you wish.

**Comment by laurafitzgerald** [comment]
@damienomurchu LGTM
